src/DataTransferObject.php

```diff
@@ -370,30 +371,43 @@ public function isDefined(string $name): bool
             return true;
         }
 
-        // Check for dot '.' notation
-        $dotPos = strpos($name, '.');
+        $sections = explode('.', $name);
+        $current = $this;
+        $defined = true;
+        $path = [];
 
-        if ($dotPos === false) {
-            if (!array_key_exists($name, $this->propertyTypes)) {
-                throw new UnknownPropertiesTypeError(static::class, [$name]);
+        while ($section = array_shift($sections)) {
+            $path[] = $section;
+            if ($current instanceof self) {
+                if (!array_key_exists($section, $current->propertyTypes)) {
+                    throw new UnknownPropertiesTypeError(static::class, [implode('.', $path)]);
+                }
+                if (array_key_exists($section, $current->properties)) {
+                    $current = $current->__get($section);
+                    continue;
+                }
+            } elseif ($current instanceof ArrayAccess) {
+                if ($current->offsetExists($section)) {
+                    $current = $current[$section];
+                    continue;
+                }
+            } elseif (is_array($current)) {
+                if (array_key_exists($section, $current)) {
+                    $current = $current[$section];
+                    continue;
+                }
+            } elseif (is_object($current)) {
+                if (property_exists($current, $section)) {
+                    $current = $current->{$section};
+                    continue;
+                }
```

Comment:
This is very much take it or leave it.

The while loop is fine, but would a foreach be more readable just on the grounds that its a more common construct?

I notice you're continuing in each block - if that's the case, is the string of elseifs better here than just separate ifs?

Reply:
Only continuing on the inner ifs not the outer if-elseif-elseif. 
Outter if -> which type
Inner if -> is this value valid for that type
continue if valid, fall through and fail if not valid.